tests/schema/tests.py

```diff
@@ -1832,3 +1833,67 @@ def test_alter_pk_with_self_referential_field(self):
+
+    @mock.patch('django.db.backends.base.schema.datetime')
+    @mock.patch('django.db.backends.base.schema.timezone')
+    def test_add_datefield_and_datetimefield_use_effective_default(self, mocked_datetime,
+                                                                   mocked_tz):
+        """
+        #25005 - effective_default() should be used for DateField,
+        DateTimeField and TimeField if auto_now or auto_add_now is set.
+        """
+        def assertDefault(field, field_name, expected_default):
+            with connection.cursor() as cursor:
+                editor.add_field(Author, field)
+                cursor.execute("SELECT %s FROM schema_author;" % field_name)
+                self.assertEqual(cursor.fetchall()[0][0], expected_default)
+
+        now = datetime.datetime(month=1, day=1, year=2000, hour=1, minute=1)
+        now_tz = datetime.datetime(month=1, day=1, year=2000, hour=1, minute=1, tzinfo=UTC())
+        mocked_datetime.now = mock.MagicMock(return_value=now)
+        mocked_tz.now = mock.MagicMock(return_value=now_tz)
+        # Create the table
+        with connection.schema_editor() as editor:
+            editor.create_model(Author)
+        # Check auto_now/auto_now_add attributes are not defined
+        columns = self.column_classes(Author)
+        self.assertNotIn("date_of_birth_auto_now", columns)
+        self.assertNotIn("date_of_birth_auto_now_add", columns)
+        self.assertNotIn("date_time_of_birth_auto_now", columns)
+        self.assertNotIn("date_time_of_birth_auto_now_add", columns)
+        self.assertNotIn("time_of_birth_auto_now", columns)
+        self.assertNotIn("time_of_birth_auto_now_add", columns)
+        # Create a row
+        Author.objects.create(name='Anonymous1')
+        # Add new date and datetime fields to ensure default will be
+        # used from effective_default
+        date_of_birth_auto_now = DateField(auto_now=True)
+        date_of_birth_auto_now.set_attributes_from_name('date_of_birth_auto_now')
+
+        date_of_birth_auto_now_add = DateField(auto_now_add=True)
+        date_of_birth_auto_now_add.set_attributes_from_name('date_of_birth_auto_now_add')
+
+        date_time_of_birth_auto_now = DateTimeField(auto_now=True)
+        date_time_of_birth_auto_now.set_attributes_from_name('date_time_of_birth_auto_now')
+
+        date_time_of_birth_auto_now_add = DateTimeField(auto_now_add=True)
+        date_time_of_birth_auto_now_add.set_attributes_from_name('date_time_of_birth_auto_now_add')
+
+        time_of_birth_auto_now = TimeField(auto_now=True)
+        time_of_birth_auto_now.set_attributes_from_name('time_of_birth_auto_now')
+
+        time_of_birth_auto_now_add = TimeField(auto_now_add=True)
+        time_of_birth_auto_now_add.set_attributes_from_name('time_of_birth_auto_now_add')
+
+        # Ensure fields was added with the right defaults
+        assertDefault(date_of_birth_auto_now, 'date_of_birth_auto_now', now.date())
```

Comment:
@shaib, on Oracle this is failing with:
```
    self.assertEqual(cursor.fetchall()[0][0], expected_default)
AssertionError: datetime.datetime(2000, 1, 1, 0, 0) != datetime.date(2000, 1, 1)
```
I didn't investigate too much but is there something obvious about Oracle datetime handling that would cause this (basically datetime is returned when date is expected)?